### PR TITLE
Introduce increment parameter in IncreaseIntegerPipe

### LINKED ISSUE #1127: IncreaseIntegerPipe with parameter value
**Reporter**
Ricardo

**Is your feature request related to a problem? Please describe.**
I want to increase an integer session key with a variable value (some sessionkey/xpathExpression), but currently I can only increment with a fixed value using the IncreaseIntegerPipe.

**Describe the solution you'd like**
To have a parameter on the IncreaseIntegerPipe with which I can update the session key.

**Describe alternatives you've considered**
Current workaround would be to use an XsltPipe with an xpathExpression and store that in a session key.

**Additional context**
N/A

### PR DESCRIPTION
**Pull request by alisihab**
Closes #1127

### CODE REVIEW

File: core/src/main/java/nl/nn/adapterframework/pipes/IncreaseIntegerPipe.java [hunk 1]
```diff
@@ -39,6 +45,7 @@
 
 	private String sessionKey=null;
 	private int increment=1;
+	private final static String INCREMENT = "increment";
```
Comment (gvanbrakel) [inline core/src/main/java/nl/nn/adapterframework/pipes/IncreaseIntegerPipe.java:48]: I would name this constant PARAMETER_INCREMENT, to indicate that it is not just the word 'increment', but meant as the name of a parameter

File: core/src/main/java/nl/nn/adapterframework/pipes/IncreaseIntegerPipe.java [hunk 2]
```diff
@@ -68,7 +89,9 @@ public String getSessionKey() {
 		return sessionKey;
 	}
 
+	@Deprecated
 	@IbisDoc({"amount to increment the value", "1"})
+	@ConfigurationWarning("Please use the parameter ["+INCREMENT+"]")
```
Comment (gvanbrakel) [inline core/src/main/java/nl/nn/adapterframework/pipes/IncreaseIntegerPipe.java:94]: It is not necessary to deprecate the attribute

File: core/src/main/java/nl/nn/adapterframework/pipes/IncreaseIntegerPipe.java [hunk 3]
```diff
@@ -52,6 +59,20 @@ public PipeRunResult doPipe(Message message, IPipeLineSession session) throws Pi
...
+			try {
+				pvl = getParameterList().getValues(message, session);
+			} catch (ParameterException e) {
+				throw new PipeRunException(this, getLogPrefix(session) + "exception extracting parameters", e);
+			}
+		}
+		ParameterValue pv = pvl.getParameterValue(INCREMENT);
+		if(pv != null && pv.getValue() != null) {
+			increment = pv.asIntegerValue(increment);
+		}
```
Comment (gvanbrakel) [inline core/src/main/java/nl/nn/adapterframework/pipes/IncreaseIntegerPipe.java:74]: The parameter value is meant to be used at runtime, so in doPipe(), and not in configure()

**Review by gvanbrakel — CHANGES_REQUESTED** [review]

File: core/src/main/java/nl/nn/adapterframework/pipes/IncreaseIntegerPipe.java (hunk 3 above)
Comment (nielsm5) [reply core/src/main/java/nl/nn/adapterframework/pipes/IncreaseIntegerPipe.java:74]: You could check (during configure) if the setter has been set/changed and if a parameter has been supplied. If so the setter will be redundant.

File: core/src/main/java/nl/nn/adapterframework/pipes/IncreaseIntegerPipe.java [hunk 4]
```diff
@@ -52,6 +57,20 @@ public PipeRunResult doPipe(Message message, IPipeLineSession session) throws Pi
...
+			try {
+				pvl = getParameterList().getValues(message, session);
+			} catch (ParameterException e) {
+				throw new PipeRunException(this, getLogPrefix(session) + "exception extracting parameters", e);
+			}
+		}
+		ParameterValue pv = pvl.getParameterValue(PARAMETER_INCREMENT);
+		if(pv != null && pv.getValue() != null) {
+			increment = pv.asIntegerValue(increment);
+		}
```
Comment (nielsm5) [inline core/src/main/java/nl/nn/adapterframework/pipes/IncreaseIntegerPipe.java:72]: The != check is also present in `asIntegerValue`.
Right now `pv.asIntegerValue(increment);` will never use the argument `increment`.
I think it's better if you either remove the != null check, or replace this argument with 0 or something?

**Review by nielsm5 — CHANGES_REQUESTED** [review]

File: core/src/main/java/nl/nn/adapterframework/pipes/IncreaseIntegerPipe.java [hunk 5]
```diff
@@ -39,20 +43,37 @@
...
+			try {
+				pvl = getParameterList().getValues(message, session);
+			} catch (ParameterException e) {
+				throw new PipeRunException(this, getLogPrefix(session) + "exception extracting parameters", e);
+			}
+		}
+		ParameterValue pv = pvl.getParameterValue(PARAMETER_INCREMENT);
+		if(pv != null) {
+			incrementBy = pv.asIntegerValue(increment);
+		}
```
Comment (gvanbrakel) [inline core/src/main/java/nl/nn/adapterframework/pipes/IncreaseIntegerPipe.java:74]: The first line will fail with NPE when parameterList == null. 
Either remove the test `` getParameterList() != null ``, or put these lines inside the guarded block

**Review by gvanbrakel — CHANGES_REQUESTED** [review]

**Review by gvanbrakel — APPROVED** [review]

File: core/src/test/java/nl/nn/adapterframework/pipes/IncreaseIntegerPipeTest.java [hunk 6]
```diff
@@ -45,4 +46,32 @@ public void testCannotIncreaseBy2AsNoSessionKey() throws Exception {
...
+		session.put(numberSession, "4");
+		Parameter inc = new Parameter();
+		inc.setName("increment");
+		inc.setValue(null);
+		pipe.addParameter(inc);
+		pipe.setSessionKey(numberSession);
+		pipe.configure();
+		doPipe(pipe, null, session);
+		assertEquals("5", session.get(numberSession));
+    }
```
Comment (nielsm5) [inline core/src/test/java/nl/nn/adapterframework/pipes/IncreaseIntegerPipeTest.java:75]: For the sake of completeness, could you also add a test with an empty String?
`testEmptyIncrementParameter()`

**Review by nielsm5 — APPROVED** [review]
